ctcp/ctcp.go

```diff
@@ -0,0 +1,30 @@
+package ctcp
+
+import "errors"
+
+// Message contains text, the text's type, and the next part of the message, if there is one
+type Message struct {
+	value       string
+	textType    TextType
+	nextMessage *Message
+}
+
+// TextType is used to specify whether an input is literal text or CTCP text
+type TextType int
+
+const (
+	// PLAIN represents literal text
+	PLAIN TextType = iota
+	// CTCP represents a CTCP request or response
+	CTCP
+)
```

Comment:
I think it would make more sense to always encode as if plain text, and decode as if CTCP, rather than passing this in. Behavior is not defined for encoding a CTCP message, is it? Same for decoding a plain text message.

Reply:
CTCP requests need to be escaped so that the text "VERSION" can be differentiated from the CTCP request of "VERSION".

Reply:
This is done with the CTCP dilimeter of `'\001'`.

Reply:
How about using an interface, which implements encode/decode?
There would be no need for a `TextType` enum if the behavior is part of the type.
I think it would be simpler that way, even if there is some repetition in struct and method declarations.

In usage, in would be the difference between
`ctcp.Message{ .., ctcp.CTCP, .. }`
and
`ctcp.CTCP{ .., .. }`

I think perhaps the package name should be related to messages, and the type should be `CTCP`, e.g. `msg.CTCP{ .., .. }` and `msg.Plain{ .., .. }`

Reply:
How would a Plaintext message be followed by a CTCP request in such a schema?

Reply:
If the implementation leaves encoding to the type of the message, then it would be:
```
// msg.go
type Encoder interface {
    Encode() string
}

type Plain struct {
    value string // or `text`
    nextMessage Encoder
}

func (p *Plain) Encode() string {
    text := p.value
    if p.nextMessage != nil {
        text += "\r\n" + p.nextMessage.Encode()
    }
    return text
}

...

// main.go
plain := &msg.Plain{ "plain text message", nil }
ctcp := &msg.CTCP{ "ctcp message", plain }

encodedText := ctcp.Encode()
```
There would be similar definitions for CTCP